### ISSUE TITLE
Module request: Nushell configuration

### ISSUE THREAD

**Issue by devurandom** [NixOS/nixpkgs#193880]
It would be nice to have a `programs.nushell` module that allows to configure nushell, similar to how `programs.fish` from [`nixos/modules/programs/fish.nix`](https://github.com/NixOS/nixpkgs/blob/nixos-22.11/nixos/modules/programs/fish.nix) does it.

As I understand it this would also be a precondition to allowing other modules, like [`nixos/modules/programs/starship.nix`](https://github.com/NixOS/nixpkgs/blob/nixos-22.11/nixos/modules/programs/starship.nix), to set up a default configuration.

**Comment by claike** [comment]
You can (and it's, I believe, the only way outside of `chsh` and manually editing `/etc/passwd`) set up per-user shells with `users.users.<name>.shell = pkgs.nushell;` ([description](https://search.nixos.org/options?channel=23.05&show=users.users.%3Cname%3E.shell&from=0&size=50&sort=relevance&type=packages&query=users.users.*.shell)). I tested it and it works perfectly well.

Also I don't think `programs.fish.enable = true;` sets fish as the default interactive shell, merely enables it in she system and allows you to configure its defaults.

Finally, it appears, that even [the developers don't recommend doing that _yet_](https://www.nushell.sh/book/default_shell.html#setting-nu-as-login-shell-linux-bsd-macos) as Nushell is not POSIX-compatible. I believe fish is also not POSIX-compatible and it isn't really a problem, so do as you please, of course  :)

I think this issue can be closed.

**Comment by claike** [comment]
However if you with to have a NixOS module that would allow you to configure Nushell's defaults declaratively you should phrase it a little different. Something like "Module request: Nushell configuration". That would also be a `9.needs: package (new)` label, not a package request, as Nushell is already in nixpkgs.

**Comment by DeeUnderscore** [comment]
`programs.fish.enable` also does stuff like add hooks to the feish config to source environmental variables set via `environment.variables` at startup. It's generally desirable to have that for a login shell, because otherwise you may end up without stuff like `PATH` or `LOCALE_ARCHIVE` set.

**Comment by tfriz343** [comment]
@devurandom Do you mind updating this issue with the info regarding claike's [last comment](https://github.com/NixOS/nixpkgs/issues/193880#issuecomment-1586376846)?

> ...Something like "Module request: Nushell configuration". That would also be a `9.needs: package (new)` label, not a package request, as Nushell is already in nixpkgs.

In order to get any movement on this, it's probably pretty important to have it titled/classified correctly, 🙂

**Comment by teto** [comment]
note that home-manager has a nushell module

**Comment by tfriz343** [comment]
Correct.  But, at least as far as I can tell, the `programs.nushell` home-manager module does not provide the functionality of loading all the required environment variables so that nushell can be used as the _default_ nixOS shell.

You can indeed set nushell as your default user shell via `users.users.<myusername>.shell = pkgs.nushell` , but you will not be able to call any or your installed packages as the PATH is not setup in the generated nushell configs (`env.nu` or `config.nu`) to allow for the calling of nix tools or installed packages.

**Comment by tfriz343** [comment]
Ah!  I didn't see the OP's updated issue description; my apologies.

Regarding my [previous comment](https://github.com/NixOS/nixpkgs/issues/193880#issuecomment-1622611744), I might be talking about a completely different issue?  That being, having the ability to configure nushell _as the default user shell_ with the correct nix environment variables and such.

@devurandom is what I described above what you're after or are you just looking to be able to run nushell _alongside_ the default shell?

If you're interested, I actually have nushell working with starship currently, thanks to the home-manager [starship module](https://github.com/nix-community/home-manager/blob/master/modules/programs/starship.nix) and also setting [programs.starship.enableNushellIntegration](https://nix-community.github.io/home-manager/options.html#opt-programs.starship.enableNushellIntegration) to true, but again, I'm not able to set nushell as my default user shell due to the `PATH`ing issue I described in my [previous comment](https://github.com/NixOS/nixpkgs/issues/193880#issuecomment-1622611744).  Currently, my workflow is, launch into the default shell (`bash` in my case) then start up `nu`, which then inherits all the environment variables from `bash`.

Happy to share my config with you if you'd like. 🙂

**Comment by devurandom** [comment]
My original goal was to have it as default shell, but as others pointed out that is not a good idea. What I understood I actually want, after using NixOS for a while now, is to be able to configure it in the manner described above.

I would be very happy to get to see configuration examples. I have not used home manager yet, but it seems to be a good way to manage my customised dotfiles.

**Comment by tfriz343** [comment]
Sure, you can find my config [here](https://gitlab.com/Lychronicity/onnix).  Happy to answer any questions.

**Comment by nyabinary** [comment]
Still would be nice to have this :3

**Comment by Septias** [comment]
While this is not a nixos module you can find a quite extensive module for home manager: https://home-manager-options.extranix.com/?query=nushell&release=release-24.05

**Comment by bugQ** [comment]
> While this is not a nixos module you can find a quite extensive module for home manager: https://home-manager-options.extranix.com/?query=nushell&release=release-24.05

Yes, this has already been brought up, along with working configurations.  There is even one such configuration on the unofficial NixOS Wiki, including a fix for the $PATH issue.  However, I think the other comments here (including myself) show that there is still sufficient interest in an actual NixOS module on par with `programs.fish`.

**Comment by jaredmontoya** [comment]
I was working on bringing support for loading environment scripts meant for bash into nushell.
more information about that here: https://github.com/nix-community/home-manager/issues/4313

But currently progress is stalled because https://github.com/NixOS/nixpkgs/pull/358950 is not getting merged due to ofborg not finishing darwin checks for more than 9 days: https://github.com/NixOS/nixpkgs/pull/358950/checks?check_run_id=33519605117.

**Comment by askeladd123** [comment]
> Yes, this has already been brought up, along with working configurations.  There is even one such configuration on the unofficial NixOS Wiki, including a fix for the $PATH issue.  However, I think the other comments here (including myself) show that there is still sufficient interest in an actual NixOS module on par with `programs.fish`.

Do you have a link for this? I couldn't find it. A configuration where Nushell is default user shell and has correct PATH / binaries.

**Comment by lychronicity** [comment]
@askeladd123  Regarding my comment prior:
> Sure, you can find my config [here](https://gitlab.com/Lychronicity/onnix). Happy to answer any questions.

This config is actually very out of date; my apologies.  I don't currently have an up-to-date config available that is public as I'm still experimenting with secret management and don't feel comfortable publishing it yet.

Regarding PATH/binaries issue I had mentioned prior (PATH not being setup/loaded correctly when setting nushell as the default user shell), this was actually an issue that was occurring when running NixOS under WSL (which I do since I was working on a Windows machine).  This has since been fixed in nix-community/NixOS-WSL#399.  After being resolved, everything has worked fine for me.  All I do to set the default shell to nushell is set the following [NixOS configuration option](https://search.nixos.org/options?channel=unstable&show=users.users.%3Cname%3E.shell&from=0&size=50&sort=relevance&type=packages&query=shell).  For example, `users.users.alice.shell = pkgs.nushell;`

As for actually configuring nushell, I use the aforementioned [module from home-manager ](https://home-manager-options.extranix.com/?query=programs.nushell&release=master).  Which in my case, my config looks like the following:
```nix
programs.nushell = {
  enable = true;
  configFile.source = ./config.nu;
  envFile.source = ./env.nu;
};
```
Hopefully this info helps.  Feel free to send an email my way if you've got further questions.  I'd like to keep further discussion of my PATH related issues and config out of this thread since my issue was-to my surprise-pretty much entirely unrelated.

**Comment by Malix-Labs** [comment]
It is tangential, but many programs are lacking a `enableNushellIntegration` because of this

Also and famously, `command-not-found` does not work with nushell (also see #532042)

However, [`installShellCompletion` does support nushell](https://github.com/NixOS/nixpkgs/blob/e711415100c53bb79af9decac2a8df18946f234d/pkgs/by-name/in/installShellFiles/setup-hook.sh#L106), weirdly enough
